# Exploration Notebook

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as pyplot
import seaborn as sns
from pathlib import Path

SEED = 42

## Dataset Overview

#### Load Data

In [2]:
train = pd.read_csv("data/raw/train.csv")
test = pd.read_csv("data/raw/test.csv")
clientes = pd.read_csv("data/raw/clientes.csv")

#### Dimensions and Structure

In [3]:
print("Train shape:", train.shape)
print("Test shape:", test.shape)
print("Clientes shape:", clientes.shape)

Train shape: (6534, 40)
Test shape: (986, 38)
Clientes shape: (1143, 13)


In [4]:
total = len(train) + len(test)

print(f"Train : {len(train):>6,} rows ({len(train) / total * 100:6.2f}%)")
print(f"Test  : {len(test):>6,} rows ({len(test) / total * 100:6.2f}%)")
print(f"Total : {total:>6,} rows (100.00%)")

Train :  6,534 rows ( 86.89%)
Test  :    986 rows ( 13.11%)
Total :  7,520 rows (100.00%)


In [5]:
print("\nTrain first 5 rows:")
display(train.head())

print("\nTest first 5 rows:")
display(test.head())

print("\nClientes first 5 rows:")
display(clientes.head())


Train first 5 rows:


,Total a Pagar,Total Bruto,Total Liquido,Total Imp_,Total Portes,Total Desc_ Global,Total Desc_ Linha,Custo,N_ Detalhes,invoice_count_current_purchase,...,customer_id,payment_type,salesperson,commercial_zone,transport_method,warehouse,document_series,purchase_date,ID,random_noise
0,558.11,510.90,453.75,104.36,0.0,0.0,57.15,153.099911,4.0,1.0,...,660784,Maybe Tomorrow Payment 10,VLOOKUP Master 9,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 61,Final Excel Warehouse 1,Ctrl Z Series 2,2024-10-29,10000000,-0.660448
1,978.80,830.02,795.77,183.03,0.0,0.0,34.25,472.943134,32.0,1.0,...,189146,Maybe Tomorrow Payment 7,VLOOKUP Master 1,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 36,Final Excel Warehouse 1,Ctrl Z Series 1,2017-07-08,10000001,0.924510
2,10.33,12.00,8.40,1.93,0.0,0.0,3.60,6.097485,1.0,1.0,...,103940,Maybe Tomorrow Payment 9,VLOOKUP Master 1,Decorative Dashboard Zone 4,NaN,Final Excel Warehouse 1,Ctrl Z Series 3,2019-03-19,10000002,-0.258730
3,856.83,696.61,696.61,160.22,0.0,0.0,0.00,408.167275,9.0,1.0,...,512927,Maybe Tomorrow Payment 7,VLOOKUP Master 4,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 71,Final Excel Warehouse 1,Ctrl Z Series 3,2018-11-10,10000003,1.448402
4,1616.10,1488.85,1313.90,302.20,0.0,0.0,174.95,485.401864,33.0,1.0,...,122356,Maybe Tomorrow Payment 12,NaN,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 61,Final Excel Warehouse 1,Ctrl Z Series 2,2021-11-30,10000004,-1.058890



Test first 5 rows:


,Total a Pagar,Total Bruto,Total Liquido,Total Imp_,Total Portes,Total Desc_ Global,Total Desc_ Linha,Custo,N_ Detalhes,invoice_count_current_purchase,...,customer_id,payment_type,salesperson,commercial_zone,transport_method,warehouse,document_series,purchase_date,ID,random_noise
0,286.37,582.05,232.82,53.55,0.0,0.0,349.23,104.409267,22,1,...,519184,Maybe Tomorrow Payment 12,VLOOKUP Master 1,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 80,Final Excel Warehouse 1,Ctrl Z Series 2,2026-08-22,10000006,0.621602
1,1949.25,1725.00,1725.00,224.25,0.0,0.0,0.00,0.000000,1,1,...,942149,Maybe Tomorrow Payment 7,VLOOKUP Master 1,Decorative Dashboard Zone 4,NaN,Final Excel Warehouse 1,Ctrl Z Series 2,2025-07-26,10000008,-1.318874
2,2738.71,4114.83,2738.71,0.00,0.0,0.0,1376.12,1019.466586,99,1,...,386654,Maybe Tomorrow Payment 4,VLOOKUP Master 1,Decorative Dashboard Zone 2,Maybe Tomorrow Delivery 54,Final Excel Warehouse 1,Ctrl Z Series 2,2026-08-22,10000010,1.251905
3,726.98,591.04,591.04,135.94,0.0,0.0,0.00,341.649008,39,2,...,588088,Maybe Tomorrow Payment 7,VLOOKUP Master 4,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 35,Final Excel Warehouse 1,Ctrl Z Series 2,2025-11-05,10000020,0.988362
4,670.76,555.49,545.33,125.43,0.0,0.0,10.16,326.715748,13,1,...,610152,Maybe Tomorrow Payment 7,VLOOKUP Master 4,Decorative Dashboard Zone 1,Maybe Tomorrow Delivery 73,Final Excel Warehouse 1,Ctrl Z Series 2,2026-07-12,10000025,0.511535



Clientes first 5 rows:


,customer_id,country_or_region,commercial_region,island,municipality,city,postal_area,postal_code,street,address,customer_type,company_name,customer_since_year
0,842570,Country CT11: Land of The Final Version,Region RG37: We Will See Later,NaN,NaN,City CY417: Missing Cell Port,NaN,Postal code PC828: #REF! Delivery Route,NaN,Address AD1120: Cell Without a Home,NaN,Excel Is Not A Database Ltd 150,2018
1,770941,Country CT9: Land of The Final Version,NaN,NaN,NaN,NaN,NaN,Postal code PC56: #REF! Delivery Route,NaN,Address AD62: Cell Without a Home,NaN,Excel Is Not A Database Ltd 1064,2021
2,827053,Country CT10: Land of The Final Version,Region RG17: We Will See Later,Island IS2: The Pivot Archipelago,Municipality MU14: Council of Last Minute Edits,NaN,Postal area PA271: Hidden Sheet Sector,Postal code PC827: #REF! Delivery Route,Street ST170: PowerPoint New Walk,Address AD449: Cell Without a Home,NaN,Excel Is Not A Database Ltd 113,2016
3,494648,Country CT10: Land of The Final Version,Region RG17: We Will See Later,Island IS2: The Pivot Archipelago,Municipality MU14: Council of Last Minute Edits,NaN,Postal area PA271: Hidden Sheet Sector,Postal code PC826: #REF! Delivery Route,NaN,Address AD454: Cell Without a Home,NaN,Excel Is Not A Database Ltd 430,2015
4,429291,Country CT10: Land of The Final Version,Region RG17: We Will See Later,Island IS2: The Pivot Archipelago,Municipality MU14: Council of Last Minute Edits,NaN,Postal area PA271: Hidden Sheet Sector,Postal code PC826: #REF! Delivery Route,NaN,Address AD453: Cell Without a Home,NaN,Excel Is Not A Database Ltd 673,2015


load data, dimensions, data types, missing values, train/test schema

In [6]:
display(test.head())

,Total a Pagar,Total Bruto,Total Liquido,Total Imp_,Total Portes,Total Desc_ Global,Total Desc_ Linha,Custo,N_ Detalhes,invoice_count_current_purchase,...,customer_id,payment_type,salesperson,commercial_zone,transport_method,warehouse,document_series,purchase_date,ID,random_noise
0,286.37,582.05,232.82,53.55,0.0,0.0,349.23,104.409267,22,1,...,519184,Maybe Tomorrow Payment 12,VLOOKUP Master 1,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 80,Final Excel Warehouse 1,Ctrl Z Series 2,2026-08-22,10000006,0.621602
1,1949.25,1725.00,1725.00,224.25,0.0,0.0,0.00,0.000000,1,1,...,942149,Maybe Tomorrow Payment 7,VLOOKUP Master 1,Decorative Dashboard Zone 4,NaN,Final Excel Warehouse 1,Ctrl Z Series 2,2025-07-26,10000008,-1.318874
2,2738.71,4114.83,2738.71,0.00,0.0,0.0,1376.12,1019.466586,99,1,...,386654,Maybe Tomorrow Payment 4,VLOOKUP Master 1,Decorative Dashboard Zone 2,Maybe Tomorrow Delivery 54,Final Excel Warehouse 1,Ctrl Z Series 2,2026-08-22,10000010,1.251905
3,726.98,591.04,591.04,135.94,0.0,0.0,0.00,341.649008,39,2,...,588088,Maybe Tomorrow Payment 7,VLOOKUP Master 4,Decorative Dashboard Zone 4,Maybe Tomorrow Delivery 35,Final Excel Warehouse 1,Ctrl Z Series 2,2025-11-05,10000020,0.988362
4,670.76,555.49,545.33,125.43,0.0,0.0,10.16,326.715748,13,1,...,610152,Maybe Tomorrow Payment 7,VLOOKUP Master 4,Decorative Dashboard Zone 1,Maybe Tomorrow Delivery 73,Final Excel Warehouse 1,Ctrl Z Series 2,2026-07-12,10000025,0.511535


In [7]:
print("\nTrain columns:")
print(train.columns.tolist())

print("\nColumns in train but not in test:")
print(set(train.columns) - set(test.columns))

print("\nColumns in test but not in train:")
print(set(test.columns) - set(train.columns))

print("\nClientes columns:")
print(clientes.columns.tolist())


Train columns:
['Total a Pagar', 'Total Bruto', 'Total Liquido', 'Total Imp_', 'Total Portes', 'Total Desc_ Global', 'Total Desc_ Linha', 'Custo', 'N_ Detalhes', 'invoice_count_current_purchase', 'repeat_purchase_90d', 'spend_next_90d', 'days_since_previous_purchase', 'customer_tenure_days', 'prior_purchase_count', 'prior_total_spend', 'prior_mean_purchase_value', 'prior_std_purchase_value', 'prior_min_purchase_value', 'prior_max_purchase_value', 'prior_mean_gap_days', 'prior_std_gap_days', 'purchase_count_30d', 'spend_30d', 'purchase_count_90d', 'spend_90d', 'purchase_count_180d', 'spend_180d', 'purchase_count_365d', 'spend_365d', 'customer_id', 'payment_type', 'salesperson', 'commercial_zone', 'transport_method', 'warehouse', 'document_series', 'purchase_date', 'ID', 'random_noise']

Columns in train but not in test:
{'repeat_purchase_90d', 'spend_next_90d'}

Columns in test but not in train:
set()

Clientes columns:
['customer_id', 'country_or_region', 'commercial_region', 'island'

#### Identify Data Types

In [8]:
data_types = pd.DataFrame({
    "data_type": train.dtypes.astype(str),
    "total_entries": len(train),
    "unique_values": train.nunique(dropna=True),
    "missing_values": train.isna().sum(),
    "missing_percentage": ((train.isna().sum() / train.shape[0]) * 100).round(2)
})

In [9]:
display(data_types)

,data_type,total_entries,unique_values,missing_values,missing_percentage
Total a Pagar,float64,6534,6122,65,0.99
Total Bruto,float64,6534,6038,65,0.99
Total Liquido,float64,6534,6093,65,0.99
Total Imp_,float64,6534,5406,65,0.99
Total Portes,float64,6534,215,65,0.99
Total Desc_ Global,float64,6534,109,65,0.99
Total Desc_ Linha,float64,6534,3519,65,0.99
Custo,float64,6534,5736,65,0.99
N_ Detalhes,float64,6534,171,65,0.99
invoice_count_current_purchase,float64,6534,5,65,0.99


Interpretation:

- A recurring group of 65 missing values

> Many purchase-related variables, such as Total a Pagar, Custo, customer_tenure_days, and spend_90d, have exactly 65 missing values.

>This suggests that the missing values may correspond to the same 65 rows. If so, a shared cause, such as incomplete records or a particular data-processing step, could explain them.

- Some features have substantial missingness

> spent_next_90d - 47.58%.

> This target-like variable has 3,109 missing values. Before deciding how to handle them, determine whether these represent unavailable future observations, customers without future purchases, or a data-generation issue.

- Historical customer features have varying missingness

> prior_std_gap_days has 22.62% missingness, while prior_mean_gap_days has 20.02% and prior_std_purchase_value has 17.03%.

>These features may be undefined for customers with too few previous purchases, rather than missing due to a data-quality problem.

##### Analyze whether the same 65 rows are missing

In [10]:
cols_65 = train.columns[train.isna().sum() == 65]
missing_65 = train[cols_65].isna()
reference = missing_65.iloc[:, 0]
same_pattern = missing_65.eq(reference, axis=0).all().all()

print("Number of columns:", len(cols_65))
print("Rows with any missing value:", missing_65.any(axis=1).sum())
print("Rows missing all selected columns:", missing_65.all(axis=1).sum())
print("Identical missing pattern:", same_pattern)

Number of columns: 24
Rows with any missing value: 65
Rows missing all selected columns: 65
Identical missing pattern: True


All 24 columns have missing values in exactly the same 65 rows.

In [11]:
cols_65 = train.columns[train.isna().sum() == 65]

affected_mask = train[cols_65].isna().all(axis=1)
affected_rows = train.loc[affected_mask]

print("Affected rows:", len(affected_rows))
print("Unique customers:", affected_rows["customer_id"].nunique())

other_cols = train.columns.difference(cols_65)

display(
    affected_rows[other_cols].head(20)
)

Affected rows: 65
Unique customers: 53


,ID,customer_id,days_since_previous_purchase,payment_type,prior_max_purchase_value,prior_mean_gap_days,prior_mean_purchase_value,prior_min_purchase_value,prior_std_gap_days,prior_std_purchase_value,repeat_purchase_90d,salesperson,spend_365d,spend_next_90d,transport_method,warehouse
393,10000457,755628,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN
484,10000560,755628,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN
582,10000678,427118,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN
583,10000679,508217,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN
748,10000870,427118,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN
762,10000890,670446,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN
817,10000955,248523,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN
837,10000980,561391,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN
839,10000983,334782,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN
954,10001116,793048,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN


- `ID` and `customer_id` are populated. These rows are not entirely empty records.

- Several groups of features are missing together.

- The target variables remain populated.

In [12]:
cols_65 = train.columns[train.isna().sum() == 65]
affected_mask = train[cols_65].isna().all(axis=1)

comparison = pd.DataFrame({
    "Affected (65 rows)": train.loc[affected_mask, "repeat_purchase_90d"],
    "Other rows": train.loc[~affected_mask, "repeat_purchase_90d"]
})

summary = pd.DataFrame({
    "Affected rows": [
        affected_mask.sum(),
        train.loc[affected_mask, "repeat_purchase_90d"].mean() * 100
    ],
    "Other rows": [
        (~affected_mask).sum(),
        train.loc[~affected_mask, "repeat_purchase_90d"].mean() * 100
    ]
}, index=["Number of rows", "Repeat purchase rate (%)"])

display(summary.round(2))

,Affected rows,Other rows
Number of rows,65.00,6469.00
Repeat purchase rate (%),50.77,55.97


##### Conclusion of the Missing Values Analysis

The missing-value analysis identified 65 transactions (0.99% of the dataset) with a shared missingness pattern across 24 features. These records retain their `repeat_purchase_90d` target, with an observed repeat-purchase rate of 50.77%, compared with 55.97% among the remaining transactions. Although the affected group exhibits a 5.20% lower repeat-purchase rate, this descriptive difference does not establish a statistically significant association. The underlying cause of the missingness should be investigated before deciding whether to exclude these observations or apply an imputation strategy.

## Data Quality and Identifiers

#### Duplicates

In [13]:
print("Exact duplicate rows in train:", train.duplicated().sum())
print("Exact duplicate rows in test:", test.duplicated().sum())


for name, df in [("train", train), ("test", test)]:
    if "ID" in df.columns:
        print(
            f"{name}: duplicate ID values =",
            df["ID"].duplicated().sum()
        )

if "ID" in train.columns and "ID" in test.columns:
    overlapping_ids = set(train["ID"].dropna()) & set(test["ID"].dropna())
    print("IDs appearing in both train and test:", len(overlapping_ids))

Exact duplicate rows in train: 0
Exact duplicate rows in test: 0
train: duplicate ID values = 0
test: duplicate ID values = 0
IDs appearing in both train and test: 0


#### Invalid Values

##### Negative Values

In [18]:
for name, df in [("train", train), ("test", test)]:
    numerical_cols = df.select_dtypes(include=[np.number]).columns
    cols_to_check = [c for c in numerical_cols if c not in ["ID", "customer_id", "random_noise"]]

    negative_dict = {}
    for col in cols_to_check:
        negative_count = (df[col] < 0).sum()
        if negative_count > 0:
            negative_dict[col] = negative_count

    if negative_dict:
        df_negatives = pd.DataFrame(list(negative_dict.items()), columns=["Column", "Negative Count"])
        df_negatives.insert(0, "DataFrame", name)
        df_negatives["Percentage"] = (df_negatives["Negative Count"] / len(df) * 100).round(2)
        display(df_negatives.sort_values(by="Negative Count", ascending=False))
    else:
        continue

,DataFrame,Column,Negative Count,Percentage
0,train,days_since_previous_purchase,16,0.24
1,train,prior_mean_purchase_value,16,0.24
2,train,prior_mean_gap_days,16,0.24
3,train,spend_365d,15,0.23


,DataFrame,Column,Negative Count,Percentage
0,test,days_since_previous_purchase,2,0.2
1,test,prior_mean_purchase_value,2,0.2
2,test,prior_mean_gap_days,2,0.2
3,test,spend_365d,2,0.2


##### Mathematical validation of Total to Pay

In [19]:
for name, df in [("train", train), ("test", test)]:
    calculate_sum = df["Total Liquido"] + df["Total Imp_"] + df["Total Portes"]
    diference = calculate_sum - df["Total a Pagar"]
    mathematical_errors = (diference.abs() > 0.05).sum()
    print(f"Mathematical errors in {name}: {mathematical_errors} rows ({mathematical_errors / len(df) * 100:.2f}%)")

    if mathematical_errors > 0:
        inconsistent_mask = diference.abs() > 0.05
        columns_facture = ['Total a Pagar', 'Total Liquido', 'Total Imp_', 'Total Portes']
        display(df.loc[inconsistent_mask, columns_facture].head(20))

Mathematical errors in train: 1053 rows (16.12%)


,Total a Pagar,Total Liquido,Total Imp_,Total Portes
6,122.83,100.43,22.40,3.05
10,195.39,158.85,36.54,4.05
20,187.18,152.18,35.00,5.90
23,73.57,69.71,3.86,5.40
27,41.38,33.64,7.74,4.89
29,586.07,476.48,109.59,4.20
31,77.74,63.83,13.91,3.35
46,265.16,215.58,49.58,5.40
48,10.24,9.37,0.87,5.60
49,37.84,34.99,2.85,4.39


Mathematical errors in test: 69 rows (7.00%)


,Total a Pagar,Total Liquido,Total Imp_,Total Portes
14,109.96,89.40,20.56,5.40
30,327.51,266.27,61.24,8.15
38,90.75,75.34,15.41,8.34
49,332.72,270.50,62.22,7.61
55,98.28,79.90,18.38,7.40
68,118.20,96.10,22.10,7.00
72,392.91,319.44,73.47,8.15
75,89.00,72.36,16.64,5.36
80,2954.24,2401.82,552.42,30.90
85,271.65,220.85,50.80,7.85


##### Excel Errors

In [20]:
excel_errors = ["#REF!", "#N/A", "#DIV/0!", "#VALUE!", "#NAME?"]
regex_pattern = "|".join(excel_errors)

text_cols_costumers = clientes.select_dtypes(include=['object', 'string']).columns
errors_excel_dict = {}
for col in text_cols_costumers:
    error_mask = clientes[col].str.contains(regex_pattern, na=False,regex=True)
    count = error_mask.sum()
    if count > 0:
        errors_excel_dict[col] = count

if errors_excel_dict:
    print("Excel errors found in 'clientes' DataFrame:")
    display(pd.DataFrame(list(errors_excel_dict.items()), columns=["Column", "Excel Error Count"]))
    example_column = list(errors_excel_dict.keys())[0]
    print(f"\nExample of rows with Excel errors in column '{example_column}':")
    display(clientes[clientes[example_column].str.contains(regex_pattern, na=False, regex=True)][[example_column]].head(20))
else:
    print("No Excel errors found in 'clientes' DataFrame.")

Excel errors found in 'clientes' DataFrame:


,Column,Excel Error Count
0,postal_code,973



Example of rows with Excel errors in column 'postal_code':


,postal_code
0,Postal code PC828: #REF! Delivery Route
1,Postal code PC56: #REF! Delivery Route
2,Postal code PC827: #REF! Delivery Route
3,Postal code PC826: #REF! Delivery Route
4,Postal code PC826: #REF! Delivery Route
5,Postal code PC825: #REF! Delivery Route
6,Postal code PC825: #REF! Delivery Route
7,Postal code PC824: #REF! Delivery Route
8,Postal code PC824: #REF! Delivery Route
9,Postal code PC823: #REF! Delivery Route


##### Validation of Dates and cumulative windows

In [22]:
for name, df in [("train", train), ("test", test)]:
    valid_dates = pd.to_datetime(df['purchase_date'].dropna())
    print(f"Interval of valid dates in {name}: {valid_dates.min().strftime('%Y-%m-%d')} to {valid_dates.max().strftime('%Y-%m-%d')}")

    spend_violations_mask = ( 
        (df['spend_30d'] > df['spend_90d']) | (df['spend_90d'] > df['spend_180d']) | (df['spend_180d'] > df['spend_365d'])
    )
    spend_violations_count = spend_violations_mask.sum()
    print(f"Inconsistencies in spending columns(ex: spend_30d > spend_90d): {spend_violations_count}")

    if spend_violations_count > 0:
        cols_spend = ['spend_30d', 'spend_90d', 'spend_180d', 'spend_365d']
        display(df.loc[spend_violations_mask, cols_spend].head(20))

    count_violations_mask = (
        (df['purchase_count_30d'] > df['purchase_count_90d']) | (df['purchase_count_90d'] > df['purchase_count_180d']) | (df['purchase_count_180d'] > df['purchase_count_365d'])
    )
    count_violations_count = count_violations_mask.sum()
    if count_violations_count > 0:
        print(f"Inconsistencies in purchase count columns(ex: purchase_count_30d > purchase_count_90d): {count_violations_count}")


Interval of valid dates in train: 2015-07-11 to 2025-02-15
Inconsistencies in spending columns(ex: spend_30d > spend_90d): 15


,spend_30d,spend_90d,spend_180d,spend_365d
263,0.00,0.00,577.67,-3456.320000
1733,0.00,0.00,264.00,-1230.720000
1744,0.00,0.00,0.00,-167.180000
1829,0.00,0.00,0.00,-753.510000
2099,2234.61,2234.61,2234.61,-2648.440000
2851,0.00,0.00,0.00,-2169.910000
3686,177.22,177.22,420.49,-1443.640000
3690,0.00,0.00,591.02,-5133.040000
3841,467.50,3204.58,9560.84,-14740.290000
3893,0.00,0.00,179.58,-4196.900000


Interval of valid dates in test: 2025-05-24 to 2026-10-06
Inconsistencies in spending columns(ex: spend_30d > spend_90d): 2


,spend_30d,spend_90d,spend_180d,spend_365d
105,0.00,974.00,974.00,-3856.64
367,837.89,837.89,837.89,-3845.79


#### ID Analysis

#### Customer_id Analysis

#### Data purchase_date

## Feature Exploration

monetary, categorical, historical, rolling frequency, rolling speed, random_noise

## Target Exploration

classification target, regression target, class imbalance, target relationships

## Modelling Readiness

correlations, train/test shift, clientes.csv, leakage, documented cleaning log